server/src/main/java/org/opensearch/cluster/metadata/CryptoMetadata.java

```diff
@@ -0,0 +1,163 @@
+        this.keyProviderName = keyProviderName;
+        this.keyProviderType = keyProviderType;
+        this.settings = settings;
+    }
+
+    /**
+     * Returns key provider name
+     *
+     * @return Key provider name
+     */
+    public String keyProviderName() {
+        return this.keyProviderName;
+    }
+
+    /**
+     * Returns key provider type
+     *
+     * @return key provider type
+     */
+    public String keyProviderType() {
+        return this.keyProviderType;
+    }
+
+    /**
+     * Returns crypto settings
+     *
+     * @return crypto settings
+     */
+    public Settings settings() {
+        return this.settings;
+    }
+
+    public CryptoMetadata(StreamInput in) throws IOException {
+        keyProviderName = in.readString();
+        keyProviderType = in.readString();
+        settings = Settings.readSettingsFromStream(in);
+    }
+
+    public static CryptoMetadata fromRequest(CryptoSettings cryptoSettings) {
+        if (cryptoSettings == null) {
+            return null;
+        }
+        return new CryptoMetadata(cryptoSettings.getKeyProviderName(), cryptoSettings.getKeyProviderType(), cryptoSettings.getSettings());
+    }
+
+    /**
+     * Writes crypto metadata to stream output
+     *
+     * @param out stream output
+     */
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(keyProviderName);
+        out.writeString(keyProviderType);
+        Settings.writeSettingsToStream(settings, out);
+    }
+
+    public static CryptoMetadata fromXContent(XContentParser parser) throws IOException {
+        XContentParser.Token token;
+        String keyProviderType = null;
+        Settings settings = null;
+        String keyProviderName = parser.currentName();
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                String currentFieldName = parser.currentName();
+                if ("key_provider_name".equals(currentFieldName)) {
+                    if (parser.nextToken() != XContentParser.Token.VALUE_STRING) {
+                        throw new OpenSearchParseException("failed to parse crypto metadata [{}], unknown type");
+                    }
+                    keyProviderName = parser.text();
+                } else if ("key_provider_type".equals(currentFieldName)) {
+                    if (parser.nextToken() != XContentParser.Token.VALUE_STRING) {
+                        throw new OpenSearchParseException("failed to parse crypto metadata [{}], unknown type");
+                    }
+                    keyProviderType = parser.text();
+                } else if ("settings".equals(currentFieldName)) {
+                    if (parser.nextToken() != XContentParser.Token.START_OBJECT) {
+                        throw new OpenSearchParseException("failed to parse crypto metadata [{}], unknown type");
```

Comment:
Not sure exactly what you mean. Each type of provider can have its  own settings and is independent of a repository. Crypto providers can be independently defined for other use cases. In remote store we have this use case defined on top of a repository and therefore, you see these settings in a put repository request.
Also, not sure which issues you are referring to during updates. As a side note, we don't allow changing crypto metadata defined over a repository. But even if we were to do that, it would have still gone via the standard cluster state update route and even there I don't see any issue as such. Regarding persisting the metadata, it is persisted the same way as repository settings. Since, crypto is a feature of repository in remote store, we persist it along with repository metadata.

Can you make your comment more clear and add details?